qiita_db/artifact.py

```diff
@@ -0,0 +1,598 @@
+                            processing_parameters.id, visibility_id,
+                            artifact_type_id, can_be_submitted_to_ebi,
+                            can_be_submitted_to_vamps, False]
+                TRN.add(sql, sql_args)
+                a_id = TRN.execute_fetchlast()
+
+                # Associate the artifact with its parents
+                sql = """INSERT INTO qiita.parent_artifact
+                            (artifact_id, parent_id)
+                         VALUES (%s, %s)"""
+                sql_args = [(a_id, p.id) for p in parents]
+                TRN.add(sql, sql_args, many=True)
+
+                instance = cls(a_id)
+            else:
+                # Create the artifact
+                sql = """INSERT INTO qiita.artifact
+                            (generated_timestamp, visibility_id,
+                             artifact_type_id, can_be_submitted_to_ebi,
+                             can_be_submitted_to_vamps, submitted_to_vamps)
+                         VALUES (%s, %s, %s, %s, %s, %s)
+                         RETURNING artifact_id"""
+                sql_args = [timestamp, visibility_id, artifact_type_id,
+                            can_be_submitted_to_ebi, can_be_submitted_to_vamps,
+                            False]
+                TRN.add(sql, sql_args)
+                a_id = TRN.execute_fetchlast()
+
+                # Associate the artifact with the prep template
+                instance = cls(a_id)
+                prep_template.artifact = instance
+                study_id = prep_template.study_id
+
+            # Associate the artifact with the study
+            sql = """INSERT INTO qiita.study_artifact (study_id, artifact_id)
+                     VALUES (%s, %s)"""
+            sql_args = [study_id, a_id]
+            TRN.add(sql, sql_args)
+
+            # Associate the artifact with its filepaths
+            fp_ids = insert_filepaths(filepaths, a_id, artifact_type,
+                                      "filepath")
+            sql = """INSERT INTO qiita.artifact_filepath
+                        (artifact_id, filepath_id)
+                     VALUES (%s, %s)"""
+            sql_args = [[a_id, fp_id] for fp_id in fp_ids]
+            TRN.add(sql, sql_args, many=True)
+            TRN.execute()
+
+        return instance
+
+    @classmethod
+    def delete(cls, artifact_id):
+        r"""Deletes an artifact from the system
+
+        Parameters
+        ----------
+        artifact_id : int
+            The artifact to be removed
+
+        Raises
+        ------
+        QiitaDBArtifactDeletionError
+            If the artifact is public
+            If the artifact has children
+            If the artifact has been analyzed
+            If the artifact has been submitted to EBI
+            If the artifact has been submitted to VAMPS
+        """
+        with TRN:
+            # This will fail if the artifact with id=artifact_id doesn't exist
```

Comment:
No, they're not stored in variables - we don't anything in memory by the object design. Thus, doing the calls `cls(artifact_id)`, `instance.visibility` and `instance.children` are accessing the database.

If we don't put these in a transaction, we can get the race condition in which a children gets created between the check and the actual deletion.